Code for OOP Assignment 1

kyptdoan opened the pull request:
Defined and implemented 3 following classes to the code base:
1. Die: to represent a die with 1 method for the die to be rolled.
2. Game: to represent a game with 3 dice, 2 methods to roll, find sum then report.
3. Testing: to implement debug tool to verify if the dice values and the sum are valid.

kyptdoan committed:
Adding methods to the base code

Adding methods to the base code to perform functions required in the assignment brief.

kyptdoan committed:
Solving data redundancy, changing method to access data

Update code solving data redundancy in Game class and accessing data from outside the class using references.

kyptdoan committed:
Correct comments using standard XML documentation

KirbyXIV commented on CMP1903_A1_2324/Program.cs:
```diff
@@ -6,14 +6,88 @@
+                Console.WriteLine("2. Type '2' if you want to quit the game.");
+
+                // check if the user inputs an integer.
+                try
```
Good Exception Handling

KirbyXIV replied:
good error handling, further expands on this on line 55

KirbyXIV commented on CMP1903_A1_2324/Program.cs:
```diff
@@ -6,14 +6,88 @@
+                }
+
+                // check if the user inputs '1' or '2' for valid options.
+                if (user_input == 1 || user_input == 2)
```
Further Handles Exception Handling

KirbyXIV commented on CMP1903_A1_2324/Game.cs:
Great XML Commenting on every file.

KirbyXIV commented on CMP1903_A1_2324/Program.cs:
```diff
@@ -6,14 +6,88 @@
+                    // option 2 is to quit the game.
+                    else
+                    {
+                        Console.WriteLine("The game ends.");
```
Allows player to exit game

KirbyXIV commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,93 @@
+        /// </summary>
+        private readonly Die die3;
+
+        /// <summary>
```
XML comments for the dies are a bit excessive

Rarisma commented on CMP1903_A1_2324/Testing.cs:
```diff
@@ -1,20 +1,55 @@
+        /// </summary>
+        public void CheckRange()
+        {
+            Debug.Assert(game.GetDie1().CurrentValue >= 1 && game.GetDie1().CurrentValue <= 6, "Invalid Die Value");
```
Could these be split up for a more comprehensive testing experience?
If you were to do it this way it should allow you to find out what exactly is wrong with a roll, i.e you would instantly know if it's too high or too low?

Rarisma reviewed (commented):
Here's what I like about your implementation:

- Great use of encapsilation
- Good use of comments, including XML ones.
- Good use of defensive programming / Error handling.
- Correct use and implementation of testing
- Meets most if not all requirements

I am unsure that your code completely conforms to the google c# style guide, for example the brace style used in your code and default c# does not conform to the c# style from my understanding.

Other than this and the other comment I made regarding your testing class; I believe your code is excellent and is easy to read. It is clear you understand the objectives of the assignment and how to use github and other relevant IDE tooling.

KirbyXIV commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,93 @@
-         * EXTRA: For extra requirements (these aren't required though), the dice rolls could be managed so that the
-         * rolls could be continous, and the totals and other statistics could be summarised for example.
-         */
+    {       
```
great XML comments

KirbyXIV commented:
Overall a good bit of code, shows you know what is asked for in the brief. good job!

Rarisma reviewed (approved):
Pr approved

kyptdoan committed:
Final changes

_ Follow the coding style guide.
_ Modified the CheckRange() method.
_ Add a method to check for user's expectation.
